src/storage/src/source/reclock/compat.rs

```diff
@@ -117,6 +117,30 @@ where
         // Allow manually simulating the scenario where the since of the remap
         // shard has advanced too far.
         fail_point!("invalid_remap_as_of");
+
+        if since.is_empty() {
+            // This can happen when, say, a source is being dropped but we on
+            // the cluster are busy and notice that only later. In those cases
+            // it can happen that we still try to render an ingestion that is
+            // not valid anymore and where the shards it uses are not valid to
+            // use anymore.
+            //
+            // This is a rare race condition and something that is expected to
+            // happen every now and then. It's not a bug in the current way of
+            // how things work.
```

Comment:
Storage has the `DroppedIds` protocol response that I assumed to exist to let the storage controller wait until all replicas have acknowledged the dropping of an object. Is that not true? Or does this race only come up in the context of 0dt upgrades?

Reply:
DroppedIds only relate to state of the storage controller, read holds are dropped right away when the collection is dropped without waiting. Right now we drop things eagerly so that a struggling, slow, broken replica does not hold up cleaning up those things, finalizing shards ,etc.